### PR TITLE
feat: support multiple btc chain config

### PR DESCRIPTION
**Pull request by ws4charlie**
# Description

- [x] Allow multiple Bitcoin chain config in zetaclient config file.
- [x] Create multiple Bitcoin observer/signer pairs when configured.

# How Has This Been Tested?

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [x] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions

## Release Notes

- **New Features**
  - Added support for stateful precompiled contracts, enhancing smart contract functionality.
  - Introduced a common importable `zetacored rpc` package for easier integration.
  - Implemented a staking precompiled contract to improve staking mechanisms.
  - Added support for restricted addresses in the Solana network for enhanced security.
  - Emission of events from the staking precompile for better transparency.
  - Expanded support for multiple Bitcoin chains, increasing compatibility.

- **Bug Fixes**
  - Improved error handling for Bitcoin chain configurations.

- **Tests**
  - Enhanced test coverage for `GetEVMConfig` and `GetBTCConfig` methods, ensuring robustness against edge cases.

### CODE REVIEW

File: zetaclient/config/types.go [hunk 1]
```diff
@@ -88,7 +88,7 @@ type Config struct {
 
 	// chain configs
 	EVMChainConfigs map[int64]EVMConfig `json:"EVMChainConfigs"`
-	BitcoinConfig   BTCConfig           `json:"BitcoinConfig"`
+	BTCChainConfigs map[int64]BTCConfig `json:"BTCChainConfigs"`
```
Comment (gartnera) [inline zetaclient/config/types.go:91]: Ideally we would preserve the old field and merge it into the new map to make the upgrade a bit less painful

Comment (ws4charlie) [reply zetaclient/config/types.go:91]: The field `BitcoinConfig` is retained by commit https://github.com/zeta-chain/node/pull/2870/commits/b1ee2eb48b4db1debb851ba77fdf68819fd2feb7 to be backward compatible

File: zetaclient/config/types.go [hunk 2]
```diff
@@ -118,12 +121,19 @@ func (c Config) GetAllEVMConfigs() map[int64]EVMConfig {
...
+func (c Config) GetBTCConfig(chainID int64) (BTCConfig, bool) {
 	c.mu.RLock()
 	defer c.mu.RUnlock()
 
-	return c.BitcoinConfig, c.BitcoinConfig != (BTCConfig{})
+	// we prefer 'BTCChainConfigs' over 'BitcoinConfig' but still fallback to be backward compatible
+	// this will allow new 'zetaclientd' binary to work with old config file
+	btcCfg, found := c.BTCChainConfigs[chainID]
+	if !found || btcCfg.Empty() {
+		btcCfg = c.BitcoinConfig
```
Comment (gartnera) [inline zetaclient/config/types.go:133]: maybe derive the chain ID from `RPCParams` so it's not accidentally mismatched?

Comment (ws4charlie) [reply zetaclient/config/types.go:133]: We actually can't figure out chain ID from the `RPCParams`. Also, the only source of truth would need to come from the RPC endpoint if there IS a way to figure out.

File: zetaclient/orchestrator/orchestrator.go [hunk 3]
```diff
@@ -410,7 +410,7 @@
 						switch {
 						case chain.IsEVM():
 							oc.ScheduleCctxEVM(ctx, zetaHeight, chainID, cctxList, ob, signer)
-						case chain.IsUTXO():
+						case chain.IsBitcoin():
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/orchestrator/orchestrator.go:413]: it's a renaming

File: zetaclient/orchestrator/bootstrap.go [hunk 4]
```diff
@@ -136,16 +136,16 @@
 			}
 
 			addSigner(chainID, signer)
-		case chain.IsUTXO():
-			cfg, found := app.Config().GetBTCConfig()
+		case chain.IsBitcoin():
+			cfg, found := app.Config().GetBTCConfig(chainID)
 			if !found {
-				logger.Std.Warn().Msgf("Unable to find UTXO config for chain %d", chainID)
+				logger.Std.Warn().Msgf("Unable to find BTC config for chain %d signer", chainID)
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/orchestrator/bootstrap.go:142]: It's just a renaming

File: zetaclient/config/types.go (hunk 2 above)
Comment (gartnera) [reply zetaclient/config/types.go:136]: Ok yeah I think this logic is fine

File: changelog.md [hunk 5]
```diff
@@ -47,6 +47,11 @@
 * [2644](https://github.com/zeta-chain/node/pull/2644) - add created_timestamp to cctx status
 * [2673](https://github.com/zeta-chain/node/pull/2673) - add relayer key importer, encryption and decryption
 * [2633](https://github.com/zeta-chain/node/pull/2633) - support for stateful precompiled contracts
+* [2788](https://github.com/zeta-chain/node/pull/2788) - add common importable zetacored rpc package
+* [2784](https://github.com/zeta-chain/node/pull/2784) - staking precompiled contract
+* [2795](https://github.com/zeta-chain/node/pull/2795) - support restricted address in Solana
+* [2861](https://github.com/zeta-chain/node/pull/2861) - emit events from staking precompile
+* [2870](https://github.com/zeta-chain/node/pull/2870) - support for multiple Bitcoin chains in the zetaclient
```
Comment (lumtis) [inline changelog.md:54]: Why do we have some many new entries?

File: zetaclient/orchestrator/orchestrator_test.go [hunk 6]
```diff
@@ -518,16 +518,15 @@ func createAppContext(t *testing.T, chainsOrParams ...any) *zctx.AppContext {
...
 
 	// Mock config
-	cfg.BitcoinConfig = config.BTCConfig{
-		RPCHost: "localhost",
-	}
-
 	for _, c := range supportedChains {
 		if chains.IsEVMChain(c.ChainId, nil) {
 			cfg.EVMChainConfigs[c.ChainId] = config.EVMConfig{Chain: c}
+		} else if chains.IsBitcoinChain(c.ChainId, nil) {
```
Comment (lumtis) [inline zetaclient/orchestrator/orchestrator_test.go:524]: Can we use switch case?

**Review by lumtis — COMMENTED** [review]
Let's wait for a review from @swift1337 for this one.

Not specifically for this PR but I think we should remove the concept of EVM and BTC configs in the future. Just keeping a list of config associated with a chain ID

File: changelog.md (hunk 5 above)
Comment (ws4charlie) [reply changelog.md]: removed accidentally included change log entries during merge https://github.com/zeta-chain/node/pull/2870/commits/582d1de93c6f8531141eb15b0a3c952fc6eba8c5

File: zetaclient/orchestrator/orchestrator_test.go (hunk 6 above)
Comment (ws4charlie) [reply zetaclient/orchestrator/orchestrator_test.go]: done in commit https://github.com/zeta-chain/node/pull/2870/commits/582d1de93c6f8531141eb15b0a3c952fc6eba8c5

**Comment by ws4charlie** [comment]
> Let's wait for a review from @swift1337 for this one.
> 
> Not specifically for this PR but I think we should remove the concept of EVM and BTC configs in the future. Just keeping a list of config associated with a chain ID

Totally agree. We'll need to a unified config struct (containing endpoint, user, password, etc.) for every external chain.

**Review by kingpinXD — APPROVED** [review]

File: cmd/zetaclientd/start_utils.go [hunk 7]
```diff
@@ -53,41 +52,34 @@ func validatePeer(seedPeer string) error {
 	return nil
 }
 
-// maskCfg sensitive fields are masked, currently only the EVM endpoints and bitcoin credentials,
+// maskCfg sensitive fields are masked, currently only the endpoints and bitcoin credentials,
 //
 //	other fields can be added.
 func maskCfg(cfg config.Config) string {
+	// Make a copy of the config
```
Comment (swift1337) [inline cmd/zetaclientd/start_utils.go:59]: Can we use a package for this? https://github.com/showa-93/go-mask 

```go
type FooBar struct {
			ABC string   `mask:"filled"` // will be masked
}
```

File: zetaclient/config/types.go [hunk 8]
```diff
@@ -91,8 +91,10 @@ type Config struct {
 
 	// chain configs
 	EVMChainConfigs map[int64]EVMConfig `json:"EVMChainConfigs"`
-	BitcoinConfig   BTCConfig           `json:"BitcoinConfig"`
-	SolanaConfig    SolanaConfig        `json:"SolanaConfig"`
+	BTCChainConfigs map[int64]BTCConfig `json:"BTCChainConfigs"`
+	// Deprecated: the 'BitcoinConfig' will be removed once the 'BTCChainConfigs' is fully adopted
+	BitcoinConfig BTCConfig    `json:"BitcoinConfig"`
```
Comment (swift1337) [inline zetaclient/config/types.go:96]: Can we make sure that `.BitcoinConfig` is not used anywhere except `GetBTCConfig(chainID)`?

File: cmd/zetaclientd/start_utils.go (hunk 7 above)
Comment (ws4charlie) [reply cmd/zetaclientd/start_utils.go:59]: looks sweet. Let's use it.

File: zetaclient/config/types.go (hunk 8 above)
Comment (ws4charlie) [reply zetaclient/config/types.go:96]: Yep. That's right.

File: cmd/zetaclientd/start_utils.go (hunk 7 above)
Comment (ws4charlie) [reply cmd/zetaclientd/start_utils.go]: https://github.com/zeta-chain/node/pull/2870/commits/1b23e969da315ab2e8738d32b5fd0e2068f99358

**Review by gartnera — APPROVED** [review]
